homeassistant/components/sonos/services.py

```diff
@@ -0,0 +1,143 @@
+"""Support to interface with Sonos players."""
+
+from __future__ import annotations
+
+import voluptuous as vol
+
+from homeassistant.components.media_player import DOMAIN as MEDIA_PLAYER_DOMAIN
+from homeassistant.const import ATTR_TIME
+from homeassistant.core import HomeAssistant, ServiceCall, SupportsResponse, callback
+from homeassistant.helpers import config_validation as cv, service
+from homeassistant.helpers.entity_platform import DATA_DOMAIN_PLATFORM_ENTITIES
+
+from .const import ATTR_QUEUE_POSITION, DOMAIN
+from .media_player import SonosMediaPlayerEntity
+from .speaker import SonosSpeaker
+
+SERVICE_SNAPSHOT = "snapshot"
+SERVICE_RESTORE = "restore"
+SERVICE_SET_TIMER = "set_sleep_timer"
+SERVICE_CLEAR_TIMER = "clear_sleep_timer"
+SERVICE_UPDATE_ALARM = "update_alarm"
+SERVICE_PLAY_QUEUE = "play_queue"
+SERVICE_REMOVE_FROM_QUEUE = "remove_from_queue"
+SERVICE_GET_QUEUE = "get_queue"
+
+ATTR_SLEEP_TIME = "sleep_time"
+ATTR_ALARM_ID = "alarm_id"
+ATTR_VOLUME = "volume"
+ATTR_ENABLED = "enabled"
+ATTR_INCLUDE_LINKED_ZONES = "include_linked_zones"
+ATTR_WITH_GROUP = "with_group"
+
+
+@callback
+def async_setup_services(hass: HomeAssistant) -> None:
+    """Register Sonos services."""
+
+    @service.verify_domain_control(hass, DOMAIN)
+    async def async_service_handle(service_call: ServiceCall) -> None:
+        """Handle dispatched services."""
+        platform_entities = hass.data.get(DATA_DOMAIN_PLATFORM_ENTITIES, {}).get(
+            (MEDIA_PLAYER_DOMAIN, DOMAIN), {}
+        )
+
+        entities = await service.async_extract_entities(
+            hass, platform_entities.values(), service_call
+        )
+
+        if not entities:
+            return
```

Comment:
Should we allow registering entity services where the service handler gets passed an iterable of entities, instead of one call per entity, so this wheel doesn't need to be reinvented in every integration which has need for it?

Integrations `amcrest`, `image_processing`, `monoprice`, `script` and `sonos` could make use of it.